File: src/main/队列/t649/PredictPartyVictory.java
package 队列.t649;

import java.util.LinkedList;

/**
 * @author: sotowang
 * @date: 2020/12/11 19:29
 */
public class PredictPartyVictory {
    public String predictPartyVictory(String senate) {
        int n = senate.length();
        LinkedList<Integer> RList = new LinkedList<>();
        LinkedList<Integer> DList = new LinkedList<>();
        for(int i=0;i<n;i++){
            if(senate.charAt(i) == 'R'){
                RList.offer(i);
            }else{
                DList.offer(i);
            }
        }
        while(!RList.isEmpty() && !DList.isEmpty()){
            int Rindex = RList.poll();
            int Dindex = DList.poll();
            if(Rindex<Dindex){
                RList.offer(Rindex+n);
            }else{
                DList.offer(Dindex+n);
            }
        }
        return RList.isEmpty()?"Dire":"Radiant";
    }
}

File: src/main/队列/t649/predictPartyVictory.go
package t649

func predictPartyVictory(senate string) string {
	var Rlist, Dlist []int
	n := len(senate)
	for i := 0; i < n; i++ {
		if senate[i] == 'R' {
			Rlist = append(Rlist, i)
		} else {
			Dlist = append(Dlist, i)
		}
	}
	for len(Rlist) > 0 && len(Dlist) > 0 {
		Rindex, Dindex := Rlist[0], Dlist[0]
		Rlist = Rlist[1:]
		Dlist = Dlist[1:]
		if Rindex < Dindex {
			Rlist = append(Rlist, Rindex+n)
		} else {
			Dlist = append(Dlist, Dindex+n)
		}
	}
	if len(Rlist) > 0 {
		return "Radiant"
	}
	return "Dire"
}
